### ISSUE TITLE
Weekly winner indicator on overall leaderboard

### ISSUE THREAD

**Issue by blreams** [blreams/collegefootballpick10#119]
Would like to highlight each weekly winner's score on the overall leaderboard (pick10/2016/results). Specifically, I'm talking about the table cell under each Wk column corresponding to that week's winner/score. Maybe color the background green with white, bold text. Something like this:

![week_winner_styling](https://cloud.githubusercontent.com/assets/5880937/19692909/b35e800c-9aa7-11e6-8959-5888787156cd.jpg)

One thing I noticed when I was looking at the generated html for this view. There are lots of duplicate id attributes on various tags. For example:

```
        <tr id="player_42">
            <td id="content">1</td>
            <td id="player-content">Chris Ruffin</td>
            <td id="highlight-content">50</td>
            <td id="content">101</td>
            <td id="content">0.633</td>

                <td id="content"><a id="week-link" href="/pick10/2016/week/1/player/42/results">8</a></td>

                <td id="content"><a id="week-link" href="/pick10/2016/week/2/player/42/results">7</a></td>

                <td id="content"><a id="week-link" href="/pick10/2016/week/3/player/42/results">7</a></td>

                <td id="content"><a id="week-link" href="/pick10/2016/week/4/player/42/results">6</a></td>

                <td id="content"><a id="week-link" href="/pick10/2016/week/5/player/42/results">3</a></td>

                <td id="content"><a id="week-link" href="/pick10/2016/week/6/player/42/results">8</a></td>

                <td id="content"><a id="week-link" href="/pick10/2016/week/7/player/42/results">6</a></td>

                <td id="content"><a id="week-link" href="/pick10/2016/week/8/player/42/results">5</a></td>

        </tr>
```

Note that each of the `<td>` tags above have the id "content" and each of the `<a>` tags have the id "week-link". That is incorrectly formed HTML. In HTML, id attributes are supposed to be unique. If you really want to refer to a group of similar tags, then you should use the class attribute, which can be shared by multiple tags.

I mention this because it would fairly easy to use css to target the weekly winner cells with a class.

**Comment by jbholden** [comment]
Is see a couple of options that might work to highlight the winners.  As far as changing the id attributes to class attributes, I think that should be tracked in a separate issue.

1.  Implement python code to get all the week winners for a particular year
2. Add a css class selector that highlights the winner as green background and white foreground

Then one of the following methods to highlight the week winner

Option 1.  pass the winners as data into the template, and then when the template is generated, add a class attribute to each player that is a winner

Option 2. pass the winners into the script and then use javascript to add a class attribute to highlight the winners (this is similar to what is done now to highlight the currently logged in players row)

**Comment by blreams** [comment]
I prefer Option 1 as it seems like fewer moving parts not having to mess with javascript. Also, I will open a separate Issue for unique id attributes.

**Comment by jbholden** [comment]
I have uploaded the branch "highlight_winner" which seems to be working.  I still need to go through the tests to make sure there are no issues.  Also, this branch currently disables using the memcache for the overall results page (for test purposes).

**Comment by jbholden** [comment]
I created a pull request for the code to highlight the winner.  Also, the pull request has a screenshot of the page for your approval.  Please review and accept the pull request or give me feedback on changes.

#128

**Comment by blreams** [comment]
I merged your pull request for highlight_winner. But there is something odd going on. I don't see any highlighted winner cells in the overall results. This was after I did a cache update. I verified in the page source that there were span tags with the attribute class="week-winner". I also verified that the week-winner class selector was defined in overall_results.css.

Here is the odd part. I was viewing page source in chrome. I tried clicking on the link for "/static/overall_results.css" and it would bring up a view of the css file, but in that view week-winner was not defined as a selector (it was missing). Which makes me think it is serving /static/ from somewhere else. I tried running python manage.py collectstatic which should have copied all the static files into whereever STATIC_ROOT is pointing in our settings.py.

I'm stumped.

**Comment by jbholden** [comment]
It shows up fine when I visit the page.  What type of browser are you using?  I am using Safari.

**Comment by blreams** [comment]
Here is what I see using Firefox (looks the same in Chrome):
![image](https://cloud.githubusercontent.com/assets/5880937/19950582/bf607b66-a12e-11e6-9ed5-a0db5e73f1f2.png)

Here is a view of the page source that shows week-winner class attribute added to one of the Player lines:
![image](https://cloud.githubusercontent.com/assets/5880937/19950632/1188ba70-a12f-11e6-9f88-ffc12c98fd8b.png)

Here is the link tag for overall_results.css in the source:
![image](https://cloud.githubusercontent.com/assets/5880937/19950672/59277eb6-a12f-11e6-8443-48d0abd5b21a.png)

When I click on that link in the source to display the underlying css file, and scroll to the bottom where the week-winner class selector was added...well, you can see it ain't there:
![image](https://cloud.githubusercontent.com/assets/5880937/19950707/97c496ae-a12f-11e6-9054-719dc1e24e62.png)

I'm starting to wonder if somehow that css is being cached by the browser. But the weird thing is I get the same behavior on 3 different computers.

**Comment by jbholden** [comment]
I ran some functional tests using firefox before I committed yesterday and it showed up there for me.  It also shows up on my iPhone.  Is there someone else we can poll to see their behavior?  There isn't some sort of CDN setup for the site?  Are you behind a proxy?

**Comment by blreams** [comment]
I can say definitively, that the problem is root caused to overall_results.css not being up-to-date with the week-winner selector. In chrome, I can bring up Inspect, add the selector manually to that file and the highlights magically appear.

So now, the question is where is /static/ getting served from. As I understand it, django has two different methods for serving static.

Method 1: When using the built-in runserver AND DEBUG=True
In this case, django has an app in INSTALLED_APPS called django.contrib.staticfiles in the settings.py file. With this app, you must set STATIC_URL and STATICFILES_DIRS variables in settings.py. You can see how they are set if you look in the file, but here is the summary:
```
BASE_DIR = os.path.dirname(os.path.dirname(__file__))
STATIC_PATH = os.path.join(BASE_DIR, 'static')
STATIC_URL = '/static/'
STATICFILES_DIRS =(STATIC_PATH,)
```
The end result of all that is static files should be served from the directory named "static" inside the base directory.

Method 2: Deployed on a webserver
In this case, django is not really involved in the actual serving of static files. But it does use some settings variables.
```
STATIC_URL = '/static/'
STATIC_ROOT = os.path.join(BASE_DIR, '../static')
```
The STATIC_ROOT variable tells django where to copy static files when you run `python manage.py collectstatic`.
STATIC_URL tells django what to use as the path for {% static ... %} template directives.

The other kicker with this method is that you have to tell pythonanywhere, 1) that '/static/' is a special path, and 2) what directory to point to for '/static/'. I've tried pointing it to static that is parallel to basedir as well as static that is inside basedir. Neither works. But it is getting some version of overall_results.css from somewhere...it just isn't the latest version.

VERY WEIRD

**Comment by blreams** [comment]
OK, I'm still not sure I know exactly what is going on...BUT...

On the pythonanywhere site, I modified the '/static/' pointer to have it point to the static folder inside collegefootballpick10 rather than the static folder in parallel with collegefootballpick10. Once I did that, I got the highlights.

OK, so what was wrong when I was running the site locally using runserver? On a lark, I tried clearing my browsing history in chrome...and voila, there were the highlights.

So, I'm actually thinking that the browser cache was the culprit all along. I don't know of a way to force a browser to reload cached static files. I did a little research on browser cache expiry and I think there are lots of moving parts when it comes to browser caching. Supposedly the server can specify in the response header parameters that tell the client how long before an item expires. But I'm not sure that pythonanywhere supports this. I suppose one option would be to move the '/static/' link to point to a different location anytime there is a css or js update that needs to be pushed.

Now that it seems to be working, I think we can close the issue.
